File: src/main/scala/ru/orderbook/v4/consumer/OrderConsumerImpl.scala
package ru.orderbook.v4.consumer

import ru.orderbook.v4.iface._
import collection.immutable.{TreeMap, HashMap}


/**
 * User: dima
 * Date: 25/03/2012
 */
class OrderConsumerImpl extends OrderConsumer {
  var log: Log = null
  var orderBooks: Map[String, OrderBook] = new HashMap().withDefaultValue( OrderBook() )

  override def startProcessing(log: Log) {
    this.log = log
    log.log(LogLevel.INFO, "Started processing orders")
  }

  override def handleEvent(action: Action, order: Order) {
    val (symbol, book) = findBookFor(order)
    val updatedBook = action match {
      case Action.ADD => book.add(order)
      case Action.EDIT => book.edit(order)
      case Action.REMOVE => book.remove(order)
    }
    orderBooks = orderBooks.updated(symbol, updatedBook)
  }

  private def findBookFor(order: Order): (String, OrderBook) = {
    if (order.getSymbol != null) (order.getSymbol, orderBooks(order.getSymbol))
    else orderBooks.find(entry => entry._2.has(order)).get
  }

  override def finishProcessing() {
    log.log(LogLevel.INFO, toPrintableString(TreeMap.empty[String, OrderBook] ++ orderBooks))
  }

  def toPrintableString(orderBookBySymbol: TreeMap[String, OrderBook]): String = {
    def bookSideToString(bookSide: BookSide) = {
      bookSide.levels.foldLeft("") { (s, levelEntry) =>
        val level = levelEntry._2
        s + "\n\tprice = " + level.price + ", size = " + level.size + ", count = " + level.count
      }
    }

    orderBookBySymbol.foldLeft("") { (s, entry) =>
      s + "\n\n" + entry._1 +
      "\nbidSide" + bookSideToString(entry._2.bidSide) +
      "\naskSide" + bookSideToString(entry._2.askSide)
    }
  }
}

File: java/src/main/java/katas/java/cmc/orderbook/v1/consumer/OrderConsumerImpl.java
package katas.java.cmc.orderbook.v1.consumer;


import katas.java.cmc.orderbook.v1.iface.*;

import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

public class OrderConsumerImpl implements OrderConsumer {

    private final Map<String, OrderBook> orderBookMap = new HashMap<String, OrderBook>();
    private final Map<Long, Order> allOrders = new HashMap<Long, Order>();
    private Log log;

    @Override
    public void startProcessing(Log log) {
        this.log = log;
        log.log(LogLevel.INFO, "Started processing orders");
    }

    @Override
    public void handleEvent(Action action, Order order) {
        switch (action) {
            case ADD:
                allOrders.put(order.getOrderId(), order);
                OrderBook orderBook = orderBookFor(order.getSymbol());
                orderBook.addOrder(order);
                break;
            case REMOVE:
                Order oldOrder = allOrders.remove(order.getOrderId());
                orderBook = orderBookFor(oldOrder.getSymbol());
                orderBook.removeOrder(oldOrder);
                break;
            case EDIT:
                oldOrder = allOrders.remove(order.getOrderId());
                orderBook = orderBookFor(oldOrder.getSymbol());
                orderBook.removeOrder(oldOrder);

                Order newOrder = new Order(
                        oldOrder.getOrderId(),
                        oldOrder.getSymbol(),
                        oldOrder.isBuy(),
                        order.getPrice(),
                        order.getQuantity()
                );
                orderBook.addOrder(newOrder);
                allOrders.put(newOrder.getOrderId(), newOrder);
                break;
        }
    }

    public OrderBook orderBookFor(String symbol) {
        OrderBook orderBook = orderBookMap.get(symbol);
        if (orderBook == null) {
            orderBook = new OrderBook(symbol);
            orderBookMap.put(symbol, orderBook);
        }
        return orderBook;
    }

    @Override
    public void finishProcessing() {
        log.log(LogLevel.INFO, toPrintableString(orderBookMap));
    }

    private static String toPrintableString(Map<String, OrderBook> orderBookMap) {
        String result = "\n";
        for (OrderBook orderBook : new TreeMap<String, OrderBook>(orderBookMap).values()) {
            result += "\n" + orderBook.asString();
        }
        return result;
    }
}
